### PR TITLE
Added custom-basic-authenticator which do Admin-Force-Password-reset-via-OTP without showing the second login page

### PR DESCRIPTION
**Pull request by GDRDABARERA**

### CODE REVIEW

File: custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.custom.authenticator.local;
+
+
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java]: Shall we remove the new line?

File: custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.net.URLEncoder;
+import java.util.HashMap;
+import java.util.Map;
+
+
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java]: Shall we remove the new line?

File: custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+    private static final long serialVersionUID = 4345354156955223654L;
+    private static final String PASSWORD_PROPERTY = "PASSWORD_PROPERTY";
+    private static final String PASSWORD_RESET_ENDPOINT = "accountrecoveryendpoint/confirmrecovery.do?";
+    private static final String CUSTOM_PASSWORD_RESET_ENDPOINT = "accountrecoveryendpoint/custom-password-reset.jsp?";
+    private static final Log log = LogFactory.getLog(BasicCustomAuthenticator.class);
+    private static String RE_CAPTCHA_USER_DOMAIN = "user-domain-recaptcha";
+
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+        String userName = request.getParameter(BasicCustomAuthenticatorConstants.USER_NAME);
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java]: Please add a new line.

File: custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public AuthenticatorFlowStatus process(HttpServletRequest request,
+                                           HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException, LogoutFailedException {
+        String userName = request.getParameter(BasicCustomAuthenticatorConstants.USER_NAME);
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java]: Please add a new line.

File: custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+                if (errorCode.equals(IdentityCoreConstants.USER_ACCOUNT_NOT_CONFIRMED_ERROR_CODE)) {
+                    retryParam = "&authFailure=true&authFailureMsg=account.confirmation.pending";
+                    String username = request.getParameter(BasicCustomAuthenticatorConstants.USER_NAME);
+                    Object domain = IdentityUtil.threadLocalProperties.get().get(RE_CAPTCHA_USER_DOMAIN);
+                    if (domain != null) {
+                        username = IdentityUtil.addDomainToName(username, domain.toString());
+                    }
+
+                    String redirectURL = response.encodeRedirectURL(loginPage + ("?" + queryParams))
+                            + BasicCustomAuthenticatorConstants.FAILED_USERNAME + URLEncoder.encode(username, BasicCustomAuthenticatorConstants.UTF_8) +
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java]: Please format these lines to follow 120 chars per line.

File: custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+                                    URLEncoder
+                                            .encode(request.getParameter(BasicCustomAuthenticatorConstants.USER_NAME)
+                                                    , BasicCustomAuthenticatorConstants.UTF_8) +
+                                    "&remainingAttempts=0";
+                        } else {
+                            redirectURL = response.encodeRedirectURL(redirectURL + ("?" + queryParams)) +
+                                    BasicCustomAuthenticatorConstants.ERROR_CODE + errorCode +
+                                    BasicCustomAuthenticatorConstants.FAILED_USERNAME +
+                                    URLEncoder
+                                            .encode(request.getParameter(BasicCustomAuthenticatorConstants.USER_NAME), BasicCustomAuthenticatorConstants.UTF_8);
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java]: Please format these lines to follow 120 chars per line.

File: custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+                        }
+                        response.sendRedirect(redirectURL);
+
+                    } else if (errorCode.equals(UserCoreConstants.ErrorCode.USER_DOES_NOT_EXIST)) {
+                        retryParam = retryParam + BasicCustomAuthenticatorConstants.ERROR_CODE + errorCode
+                                + BasicCustomAuthenticatorConstants.FAILED_USERNAME + URLEncoder
+                                .encode(request.getParameter(BasicCustomAuthenticatorConstants.USER_NAME), BasicCustomAuthenticatorConstants.UTF_8);
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java]: Please format these lines to follow 120 chars per line.

File: custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+                        response.sendRedirect(response.encodeRedirectURL(loginPage + ("?" + queryParams))
+                                + BasicCustomAuthenticatorConstants.AUTHENTICATORS + getName() + ":" +
+                                BasicCustomAuthenticatorConstants.LOCAL + retryParam);
+                    } else if (errorCode.equals(IdentityCoreConstants.ADMIN_FORCED_USER_PASSWORD_RESET_VIA_OTP_MISMATCHED_ERROR_CODE)) {
+                        retryParam = "&authFailure=true&authFailureMsg=login.fail.message";
+                        String redirectURL = response.encodeRedirectURL(loginPage + ("?" + queryParams)) +
+                                BasicCustomAuthenticatorConstants.FAILED_USERNAME + URLEncoder.encode(request.getParameter(
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java]: Please format these lines to follow 120 chars per line.

File: custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+                                BasicCustomAuthenticatorConstants.USER_NAME), BasicCustomAuthenticatorConstants.UTF_8) + BasicCustomAuthenticatorConstants.ERROR_CODE + errorCode
+                                + BasicCustomAuthenticatorConstants.AUTHENTICATORS + getName() + ":" + BasicCustomAuthenticatorConstants.LOCAL + retryParam;
+                        response.sendRedirect(redirectURL);
+                    } else {
+                        retryParam = retryParam + BasicCustomAuthenticatorConstants.ERROR_CODE + errorCode
+                                + BasicCustomAuthenticatorConstants.FAILED_USERNAME + URLEncoder
+                                .encode(request.getParameter(BasicCustomAuthenticatorConstants.USER_NAME), BasicCustomAuthenticatorConstants.UTF_8);
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java]: Please format these lines to follow 120 chars per line.

File: custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+    @Override
+    public String getFriendlyName() {
+        return BasicCustomAuthenticatorConstants.AUTHENTICATOR_FRIENDLY_NAME;
+    }
+
+    @Override
+    public String getName() {
+        return BasicCustomAuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+}
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java]: Add a new line pls.

File: custom-basic-authenticator-for-ForcePasswordOTP/resources/accountrecoveryendpoint/custom-password-reset-complete.jsp [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+
+        Property tenantProperty = new Property();
+        tenantProperty.setKey(IdentityManagementEndpointConstants.TENANT_DOMAIN);
+        if (tenantDomain == null) {
+            tenantDomain = IdentityManagementEndpointConstants.SUPER_TENANT;
+        }
+        tenantProperty.setValue(URLEncoder.encode(tenantDomain, "UTF-8"));
+        properties.add(tenantProperty);
+
+
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/resources/accountrecoveryendpoint/custom-password-reset-complete.jsp]: Please remove additional new lines.

File: custom-basic-authenticator-for-ForcePasswordOTP/resources/accountrecoveryendpoint/custom-password-reset.jsp [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+    String errorMsg = IdentityManagementEndpointUtil.getStringValue(request.getAttribute("errorMsg"));
+    String confirmationKey = request.getParameter("confirmation");
+    String callback = request.getParameter("callback");
+    String tenantDomain = request.getParameter(IdentityManagementEndpointConstants.TENANT_DOMAIN);
+    String username = request.getParameter("username");
+    String relayingParty = request.getParameter("relyingParty");
+    String sp = request.getParameter("sp");
+    String commonAuthCallerPath = request.getParameter("commonAuthCallerPath");
+    String sessionDataKey = request.getParameter("sessionDataKey");
+
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/resources/accountrecoveryendpoint/custom-password-reset.jsp]: Please remove additional new lines.

File: custom-basic-authenticator-for-ForcePasswordOTP/resources/accountrecoveryendpoint/custom-password-reset.jsp [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+                            %>
+                            <div>
+                                <input type="hidden" name="sessionDataKey"
+                                       value="<%=Encode.forHtmlAttribute(sessionDataKey)%>"/>
+                            </div>
+                                <%
+                                }
+                            %>
+
+
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/resources/accountrecoveryendpoint/custom-password-reset.jsp]: Please remove additional new lines.

File: custom-basic-authenticator-for-ForcePasswordOTP/resources/authenticationendpoint/login.jsp [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+        List<String> localAuthenticatorNames = new ArrayList<String>();
+
+        if (idpAuthenticatorMapping != null && idpAuthenticatorMapping.get(Constants.RESIDENT_IDP_RESERVED_NAME) != null) {
+            String authList = idpAuthenticatorMapping.get(Constants.RESIDENT_IDP_RESERVED_NAME);
+            if (authList != null) {
+                localAuthenticatorNames = Arrays.asList(authList.split(","));
+            }
+        }
+
+
```
Comment (vihanga-liyanage) [inline custom-basic-authenticator-for-ForcePasswordOTP/resources/authenticationendpoint/login.jsp]: Please remove additional new lines.

File: custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java (hunk 3 above)
Comment (GDRDABARERA) [reply custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java]: added

File: custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java (hunk 4 above)
Comment (GDRDABARERA) [reply custom-basic-authenticator-for-ForcePasswordOTP/org.wso2.custom.authenticator.local/src/main/java/org/wso2/custom/authenticator/local/BasicCustomAuthenticator.java]: added

File: custom-basic-authenticator-for-ForcePasswordOTP/resources/accountrecoveryendpoint/custom-password-reset.jsp (hunk 12 above)
Comment (GDRDABARERA) [reply custom-basic-authenticator-for-ForcePasswordOTP/resources/accountrecoveryendpoint/custom-password-reset.jsp]: removed

**Pull request merged**